File: backend/middleware/auth.middleware.js
const jwt = require("jsonwebtoken");

exports.verificarToken = (req, res, next) => {

    const authHeader = req.headers.authorization;

    if (!authHeader) {
        return res.status(401).json({ mensaje: "Token requerido" });
    }

    const token = authHeader.split(" ")[1];

    if (!token) {
        return res.status(401).json({ mensaje: "Token inválido" });
    }

    try {

        const decoded = jwt.verify(token, process.env.JWT_SECRET);

        req.usuario = decoded;

        next();

    } catch (error) {

        console.error("Error verificando token:", error);

        return res.status(403).json({ mensaje: "Token inválido o expirado" });

    }

};

exports.verificarAdmin = (req, res, next) => {

    if (req.usuario.rol !== "admin") {
        return res.status(403).json({ mensaje: "Acceso solo para administradores" });
    }

    next();

};

File: backend/middleware/auth_middleware.py
from fastapi import Depends, HTTPException
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from jose import jwt
import os

SECRET_KEY = os.getenv("JWT_SECRET")
ALGORITHM = "HS256"

security = HTTPBearer()

def verificarToken(credentials: HTTPAuthorizationCredentials = Depends(security)):
    try:
        token = credentials.credentials
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        return payload
    except:
        raise HTTPException(status_code=401, detail="Token inválido")


def verificarAdmin(usuario=Depends(verificarToken)):
    if usuario.get("rol") != "admin":
        raise HTTPException(status_code=403, detail="No autorizado")
    return usuario